tool/docker_linux/Dockerfile

```diff
@@ -0,0 +1,140 @@
+  python-pip \
+  python-setuptools \
+  ruby-full \
+  nodejs \
+  lib32stdc++6 \
+  libstdc++6 \
+  libglu1-mesa \
+  build-essential \
+  default-jdk-headless
+
+RUN pip install wheel
+RUN gpg --keyserver hkp://keys.gnupg.net --recv-keys 409B6B1796C275462A1703113804BB82D39DC0E3 7D2BAF1CF37B13E2069D6956105BD0E739499BDB
+RUN curl -sSL https://get.rvm.io | bash -s stable
+ENV PATH="/usr/local/rvm/bin:${PATH}"
+RUN usermod -a -G rvm root
+RUN rvm install 2.4.3
+
+# Install the Android SDK Dependency.
+ENV ANDROID_SDK_URL="https://dl.google.com/android/repository/sdk-tools-linux-4333796.zip"
+ENV ANDROID_TOOLS_ROOT="/opt/android_sdk"
+RUN mkdir -p "${ANDROID_TOOLS_ROOT}"
+RUN mkdir -p ~/.android
+# Silence warning.
+RUN touch ~/.android/repositories.cfg
+ENV ANDROID_SDK_ARCHIVE="${ANDROID_TOOLS_ROOT}/archive"
+RUN wget --progress=dot:giga "${ANDROID_SDK_URL}" -O "${ANDROID_SDK_ARCHIVE}"
+RUN unzip -q -d "${ANDROID_TOOLS_ROOT}" "${ANDROID_SDK_ARCHIVE}"
+# Suppressing output of sdkmanager to keep log size down
+# (it prints install progress WAY too often).
+RUN yes "y" | "${ANDROID_TOOLS_ROOT}/tools/bin/sdkmanager" "tools" > /dev/null
+RUN yes "y" | "${ANDROID_TOOLS_ROOT}/tools/bin/sdkmanager" "build-tools;28.0.0" > /dev/null
+RUN yes "y" | "${ANDROID_TOOLS_ROOT}/tools/bin/sdkmanager" "platforms;android-28" > /dev/null
+RUN yes "y" | "${ANDROID_TOOLS_ROOT}/tools/bin/sdkmanager" "platform-tools" > /dev/null
+RUN yes "y" | "${ANDROID_TOOLS_ROOT}/tools/bin/sdkmanager" "extras;android;m2repository" > /dev/null
+RUN yes "y" | "${ANDROID_TOOLS_ROOT}/tools/bin/sdkmanager" "extras;google;m2repository" > /dev/null
+RUN yes "y" | "${ANDROID_TOOLS_ROOT}/tools/bin/sdkmanager" "patcher;v4" > /dev/null
+RUN rm "${ANDROID_SDK_ARCHIVE}"
+ENV PATH="${ANDROID_TOOLS_ROOT}/tools/bin:${ANDROID_TOOLS_ROOT}/tools:${PATH}"
+# Silence warnings when accepting android licenses.
+RUN mkdir -p ~/.android
+RUN touch ~/.android/repositories.cfg
+
+# Setup gradle
+ENV GRADLE_ROOT="/opt/gradle"
+RUN mkdir -p "${GRADLE_ROOT}"
+ENV GRADLE_ARCHIVE="${GRADLE_ROOT}/gradle.zip"
+ENV GRADLE_URL="http://services.gradle.org/distributions/gradle-4.4-bin.zip"
+RUN wget --progress=dot:giga "$GRADLE_URL" -O "${GRADLE_ARCHIVE}"
+RUN unzip -q -d "${GRADLE_ROOT}" "${GRADLE_ARCHIVE}"
+ENV PATH="$GRADLE_ROOT/bin:$PATH"
+
+# Add npm to path.
+ENV PATH="/usr/bin:${PATH}"
+RUN dpkg-query -L nodejs
+# Install Firebase tools.
+RUN npm install --global firebase-tools@3.19.3
+
+# Set locale to en_US.UTF-8
+RUN dpkg-reconfigure locales
+RUN locale-gen "en_US" "en_US.UTF-8" "C.UTF-8" && /usr/sbin/update-locale LANG="C.UTF-8"
+RUN echo "en_US.UTF-8 UTF-8" >> /etc/locale.gen && locale-gen
+ENV LC_ALL="C.UTF-8"
+ENV LANG="en_US.UTF-8"
+ENV LANGUAGE="en_US.UTF-8"
+
+# Install the bundler so we can install Ruby bundles.
+RUN gem install bundler
+
+# Installing Ruby bundles.
+ENV BUNDLE_ROOT="/opt/bundler"
+ENV BUNDLE_GEMFILE="$BUNDLE_ROOT/Gemfile"
+RUN mkdir --parents "$BUNDLE_ROOT"
+ADD Gemfile "$BUNDLE_GEMFILE"
```

Comment:
No, files added to Docker images must be below the Dockerfile.  I could move the Docker file to the top level, but that's too bad: it feels like leaking the CI config into the website. Can they install gems from this directory?

Reply:
What do you do when multiple platforms are used? Duplicate the Gemfiles?

Can we symlink to Gemfiles that are elsewhere?
If not, and they have to be here, then we'll adapt.

Reply:
OK, I'll try symlinking them.

Reply:
OK, well, I tried symlinking them both directions, and the Docker build fails if they're symlinks in the docker directory (because Docker is smart enough to tell that they're pointing outside of it's subtree), and the jekyll build fails if they're symlinks at the top level (because they get copied into _site, and then the relative links don't work anymore).

Is it OK to leave them in the docker dir?